lib/Interpreter/CppInterOp.cpp

```diff
@@ -1027,61 +1056,78 @@ namespace Cpp {
-        continue;
-#else // CLANG_REPL
-      if (func->getMinRequiredArguments() > arg_types.size())
-        continue;
+    cling::Interpreter::PushTransactionRAII RAII(&getInterp());
 #endif
 
-      // TODO(aaronj0) : first score based on the type similarity before forcing
-      // instantiation.
-
-      TCppFunction_t instantiated =
-          InstantiateTemplate(candidate, arg_types.data(), arg_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // Force the instantiation with template params in case of no args
-      // maybe steer instantiation better with arg set returned from
-      // TemplateProxy?
-      instantiated = InstantiateTemplate(candidate, explicit_types.data(),
-                                          explicit_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // join explicit and arg_types
-      std::vector<TemplateArgInfo> total_arg_set;
-      total_arg_set.reserve(explicit_types.size() + arg_types.size());
-      total_arg_set.insert(total_arg_set.end(), explicit_types.begin(),
-                           explicit_types.end());
-      total_arg_set.insert(total_arg_set.end(), arg_types.begin(),
-                           arg_types.end());
-
-      instantiated = InstantiateTemplate(candidate, total_arg_set.data(),
-                                         total_arg_set.size());
-      if (instantiated)
-        return instantiated;
+    llvm::SmallVector<Expr*> Args;
+    for (auto i : arg_types) {
+      QualType Type = QualType::getFromOpaquePtr(i.m_Type);
+      ExprValueKind ExprKind = ExprValueKind::VK_PRValue;
+      if (Type->isReferenceType())
+        ExprKind = ExprValueKind::VK_LValue;
+      Args.push_back(new (getSema().getASTContext())
+                         OpaqueValueExpr(SourceLocation::getFromRawEncoding(1),
+                                         Type.getNonReferenceType(), ExprKind));
     }
-    return nullptr;
+
+    // Create a list of template arguments.
+    llvm::SmallVector<TemplateArgument> TemplateArgs;
+    TemplateArgs.reserve(explicit_types.size());
+    for (auto explicit_type : explicit_types) {
+      QualType ArgTy = QualType::getFromOpaquePtr(explicit_type.m_Type);
+      if (explicit_type.m_IntegralValue) {
+        // We have a non-type template parameter. Create an integral value from
+        // the string representation.
+        auto Res = llvm::APSInt(explicit_type.m_IntegralValue);
+        Res = Res.extOrTrunc(C.getIntWidth(ArgTy));
+        TemplateArgs.push_back(TemplateArgument(C, Res, ArgTy));
+      } else {
+        TemplateArgs.push_back(ArgTy);
+      }
+    }
+
+    TemplateArgumentListInfo TLI{};
+    for (auto TA : TemplateArgs)
+      TLI.addArgument(
+          S.getTrivialTemplateArgumentLoc(TA, QualType(), SourceLocation()));
+
+    OverloadCandidateSet Overloads(
+        SourceLocation(), OverloadCandidateSet::CandidateSetKind::CSK_Normal);
+    for (void* i : candidates) {
```

Comment:
Why not using [Sema::AddFunctionCandidates](https://clang.llvm.org/doxygen/classclang_1_1Sema.html#abb87d8a7ae2840db51030e162f3018dc) instead of this loop?

Reply:
Maybe that assertion is coming from an interface on our end that needs fixing?

Reply:
> We will still require a loop to populate UnresolvedSetImpl from candidates.

It seems to have an `append` operator taking a range...

Reply:
> It seems to have an append operator taking a range...

The iterator is a [`UnresolvedSetIterator`](https://clang.llvm.org/doxygen/classclang_1_1UnresolvedSetIterator.html)(reference: https://clang.llvm.org/doxygen/classclang_1_1UnresolvedSetImpl.html#ad1e4dde3cd343da5f9918eb0caedd0a7)

---

> Maybe that assertion is coming from an interface on our end that needs fixing?

I have been debugging this. I am not able to detect the cause. Our interfaces are being used correctly, as far as I can see.